core/src/main/scala/chisel3/Placeholder.scala

```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3
+
+import chisel3.experimental.SourceInfo
+import chisel3.internal.Builder
+import chisel3.internal.firrtl.ir
+
+/** A [[Placeholder]] is an _advanced_ API for Chisel generators to generate
+  * additional hardware at a different point in the design.
+  *
+  * For example, this can be used to add a wire _before_ another wire:
+  * {{{
+  * val placeholder = new Placeholder()
+  * val a = Wire(Bool())
+  * val b = placeholder.append {
+  *   Wire(Bool())
+  * }
+  * }}}
+  *
+  * This will generate the following FIRRTL where `b` is declared _before_ `a`:
+  * {{{
+  * wire b : UInt<1>
+  * wire a : UInt<1>
+  * }}}
+  */
+class Placeholder()(implicit sourceInfo: SourceInfo) {
```

Comment:
Mentioned to you offline, but I would rather this be a private API, at least initially. I admit I do have use cases for this in private code using Chisel, but I would rather box this up and put a bow on it as some sort of "Chisel Lazy Val" API rather than a mutable box that you can keep attaching things to.

Reply:
Made private in the rebase. Good thinking. If we ever do want such a footgun of an API made public, we can first figure out how to do use-before-def checking with changes to the Chisel Builder (or its evolved Scala3 form).